### PR TITLE
average_price edges; pin current IOC/GTD behavior at the level (#79)

### LINKED ISSUE #79: Add TIF boundary, GTD/Day expiry, and average_price edge tests
## Overview

Time-in-force boundary behavior and `average_price` numeric edges are untested,
and the existing TIF tests are misnamed (they assert "behaves like standard
orders"). Add named boundary tests for IOC discard, FOK kill-side, GTD/Day
expiry, and `average_price` precision/NaN guards. The IOC/FOK/expiry framing
depends on the taker-semantics design decision (see related issue).

## Context

- `src/price_level/tests/level.rs:1078` (`test_match_fill_or_kill_order`) tests
  only the fill side; no kill-side test (taker qty > resting depth must leave
  zero trades + queue untouched).
- `:1096` (`test_match_immediate_or_cancel_order`) never exercises the discard
  boundary (remainder discarded, taker not enqueued, `remaining > 0`).
- `:1114` (`test_match_good_till_date_order`) matches a GTD order as if it never
  expires; `TimeInForce::is_expired` is unit-tested in isolation
  (`src/orders/tests/time_in_force.rs:26,35`) but never consulted by the level.
- `src/execution/match_result.rs:135` `average_price` divides f64/f64; tested
  only for one tidy value + the zero case (`tests/match_result.rs:121-134`); no
  large-value precision test, no `Ok(None)` on zero-executed from real
  `match_order` output.
- Rules: `rules/global_rules.md` "Matching" (IOC/FOK exactness, Gtd/Day expiry;
  test the filled/killed boundary), "Arithmetic" (guard f64 NaN/Inf),
  "Testing" (naming).

## Tasks

- [ ] Named IOC test: `test_match_order_partial_fill_ioc_discards_remainder`.
- [ ] Named FOK kill-side test (all-or-nothing, no partial state left behind).
- [ ] GTD/Day: test that expired makers are skipped, or pin documented caller
      responsibility (per the design decision).
- [ ] `average_price` precision test with large value/qty + zero-executed →
      `Ok(None)` on engine output.

## Acceptance criteria

- [ ] TIF boundary, expiry, and `average_price` edge tests present and passing
      (consistent with the taker-semantics decision).
- [ ] Misleading TIF test names corrected.
- [ ] `cargo clippy --all-targets --all-features -- -D warnings` clean
- [ ] `cargo fmt --all --check` clean
- [ ] `cargo test` clean
- [ ] `cargo build --release` clean

## References

- `src/price_level/tests/level.rs:1078,1096,1114`,
  `src/execution/match_result.rs:135`, `src/orders/tests/time_in_force.rs`
- Related: taker-semantics design issue
- `rules/global_rules.md` (Matching, Arithmetic, Testing)

### PR DESCRIPTION
**Pull request by joaquinbejar**
## Summary

Adds the non-gated parts of the TIF / expiry / `average_price` test coverage:
`average_price` edge cases, a pin of the level's de-facto IOC behavior, and a
test + doc note that time-in-force expiry is the caller's responsibility. The
parts that need taker-TIF enforcement (FOK all-or-nothing, true IOC discard) are
deferred to the gated #65.

## Changes

- **`average_price` edges** (`match_result_trade.rs`): zero-executed → `Ok(None)`;
  exact small values; the `f64` precision-loss case; and a never-`NaN`/`Inf`
  sweep over valid inputs.
- **IOC de-facto pin** (`level.rs` tests):
  `test_match_order_taker_exceeds_depth_fills_available_and_reports_remainder` —
  a larger-than-depth taker fills available, leaves `remaining_quantity() > 0`,
  `is_complete() == false`, and is never enqueued.
- **GTD/Day**: `test_match_order_does_not_enforce_gtd_maker_expiry` — a Gtd maker
  with a past expiry still matches; added a `match_order` rustdoc note that TIF
  expiry is not enforced at the level (caller/order-book responsibility).
- Clarifying comments on the existing `test_match_fill_or_kill_order` /
  `test_match_immediate_or_cancel_order` (maker TIF not enforced; matches like
  Gtc).

## Technical decisions

- **`average_price` precision (no bug):** for `price = 2^53 + 1, quantity = 1`
  the exact average is `9_007_199_254_740_993` but the `f64` division yields
  `…992.0` (off by 1) — `2^53 + 1` isn't representable in an f64 mantissa. The
  matching path keeps exact integer math; only `average_price` analytics drop to
  `f64`. Never `NaN`/`Inf` for valid inputs. Tests assert closeness + finiteness
  and document the limitation.
- **Deferred to #65:** FOK all-or-nothing "kill" and IOC discard *enforcement* —
  `match_order` takes no taker `TimeInForce` and never enqueues the taker, so it
  can't distinguish fill-or-kill from fill-what-you-can. The IOC test pins only
  the observable behavior the engine already has.

## Testing

- [x] `make pre-push` clean

`cargo test`: 380 (unit) + 1 (integration) + 9 (doctests) passed, 0 failed.
`cargo clippy --all-targets --all-features -- -D warnings`, `cargo fmt --all
--check`, `cargo build --release`: all clean.

## Checklist

- [x] `average_price` zero / precision / NaN-guard edges covered
- [x] Current IOC + GTD/Day behavior pinned and documented (expiry = caller-side)
- [x] FOK/IOC *enforcement* explicitly deferred to #65 (no faked behavior)
- [x] Tests + one rustdoc note only; no behavior change; no new deps

Closes #79

### CODE REVIEW

File: src/execution/tests/match_result_trade.rs [hunk 1]
```diff
@@ -86,4 +86,143 @@ mod tests {
...
     }
+
+    // ----- average_price edge cases -----
+    //
+    // `average_price()` returns `Result<Option<f64>, PriceLevelError>` and
+    // computes `executed_value as f64 / executed_quantity as f64`, guarding
+    // the zero-quantity case so it never divides by zero. These tests pin the
+    // zero-quantity, precision-loss, and never-NaN/Inf behavior. The
+    // happy-path average (`test_average_price`) is verified via real
+    // `match_order` output in the price-level tests.
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply src/execution/tests/match_result_trade.rs]: Fixed the comment. The happy-path exact average is covered by test_average_price_exact_small_values_is_precise in this same module, not the price-level tests.

File: src/execution/tests/match_result_trade.rs [hunk 2]
```diff
@@ -86,4 +86,143 @@ mod tests {
...
+                // Precision genuinely drifts: the f64 average differs from the
+                // exact integer average by exactly 1 here. Assert it is close,
+                // not equal.
+                let exact = EXACT_INT_AVG as f64;
+                let abs_err = (avg - exact).abs();
+                assert!(
+                    abs_err <= 2.0,
+                    "f64 average should be within tolerance of the exact \
+                     integer average; abs_err = {abs_err}"
+                );
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply src/execution/tests/match_result_trade.rs]: Right, EXACT_INT_AVG as f64 rounded the same way so abs_err was always 0. The test now converts the f64 result back to u128 and asserts it falls short of the exact integer average by exactly 1, which actually demonstrates the loss while still pinning the observed f64 value.

**Comment by joaquinbejar** [comment]
Thanks, both fixed. The off-by-1 is now shown in integer space (the f64 result converted back to u128 is short by exactly 1), and the module comment points to the right coverage.
